### ISSUE TITLE
ruby: jit support

### ISSUE THREAD

**Issue by Mindavi** [NixOS/nixpkgs#125260]
In this PR: https://github.com/NixOS/nixpkgs/pull/119506 JIT support was disabled by default for ruby. This was done on the basis of the closure size, which grows significantly due to the inclusion of LLVM.

There are some questions here:

- Do we prioritize closure size or features?
- Is there a middle ground that ensures we have binary substitution for both camps?

I understand that some people want to keep closure size down, while others want all the features.

In the case of @dasJ, it seems like gitlab performance has decreased to an unacceptable level, so for them it's very desirable to have JIT support enabled.

What would be the 'right' decision here? Can we cater for both while having a sane default, or do we have to compromise and choose who has to 'rebuild the (ruby) world'?

As far as I've seen within nixpkgs, typically there's a tendency to enable most or all features. I think that's most common within package management in general. But I might be mistaken and we might have some guidelines on this. I haven't checked, please let me know if there's some guidance.

**Comment by dasJ** [comment]
Our solution was something like this in our overlay:
```nix
self: super: {
  ruby_2_7_jit = super.ruby_2_7.override { jitSupport = true; };

  gitaly = super.callPackage ./gitlab/gitaly { ruby = self.ruby_2_7_jit; };
  gitlab = super.callPackage ./gitlab { ruby = self.ruby_2_7_jit; };
  gitlab-ee = super.callPackage ./gitlab { gitlabEnterprise = true; ruby = self.ruby_2_7_jit; };
  gitlab-shell = super.callPackage ./gitlab/gitlab-shell { ruby = self.ruby_2_7_jit; };
}
```

Maybe it's a good idea to do that to applications like GitLab where the closure size does not matter (it has 1.6GB for us). Also cc @ajs124

**Comment by doronbehar** [comment]
I think the majority of the users don't need JIT support for the few Ruby applications they use. A good example for that is what made me notice [the issue](https://github.com/NixOS/nixpkgs/issues/89445) and open https://github.com/NixOS/nixpkgs/pull/90151 initially. My only real ruby package on my system is `asciidoctor` which is popular a bit I think and IMO the average user will prefer the reduced closure size in exchange for a possible performance increase.

I upvote @dasJ's idea, we won't use `recurseIntoAttrs` so it won't be a big burden for Hydra IMO.

**Comment by dasJ** [comment]
Looks like discourse also uses rails and would therefore benefit from the jit

**Comment by daniels** [comment]
This is old, but I just noticed how large the ruby closure in nixpkgs has become from ruby_3_2:

```console
$ # Note: nixpkgs-stable is my current pin of nixpkgs-24.05-darwin
$ nix path-info --closure-size --human-readable nixpkgs-stable#ruby_3_2
/nix/store/5isal3br0dg8anzaj3zkgb8kclsv3jcj-ruby-3.2.4   917.1M
$ nix path-info --closure-size --human-readable nixpkgs-stable#ruby_3_1
/nix/store/sl4zvdgpn2xpjncgllqdnnah90fdaix9-ruby-3.1.6   101.9M
```

Before I've used nix ruby only on my developer machine. But my new use case is that I want to build containers using nix dockerTools, and the size of the final image with ruby 3.2 is 3x larger with (y)jit-enabled vs without jit. One of the selling points of building with dockerTools is that it's easy to build really small images ...

I'd love to have a choice of jit/non-jit rubies available from the binary cache.

**Comment by doronbehar** [comment]
But @daniels just to make sure, are you capable of reducing the closure size of `ruby_3_2` if you disable JIT? Your complain is only about the lack of this variant in the binary cache?

**Comment by daniels** [comment]
Yes, @doronbehar, I've disabled with `yjitSupport = false`, and it works well enough. But yes, I wish this was available in the binary cache. It's not like I believe the binary cache should provide every conceivable configuration, but as the difference is this huge I think it might be called for in this case.

(For me the required rebuild hits three times - I build for aarch64-darwin, aarch64-linux, and x86_64-linux on my machine.)
